driver/modern_bpf/helpers/store/auxmap_store_params.h

```diff
@@ -267,15 +272,116 @@ static __always_inline void auxmap__store_u64_param(struct auxiliary_map *auxmap
 {
-	u16 charbuf_len = push__charbuf(auxmap->data, &auxmap->payload_pos, charbuf_pointer, MAX_PARAM_SIZE);
+	u16 charbuf_len = push__charbuf(auxmap->data, &auxmap->payload_pos, charbuf_pointer, MAX_PARAM_SIZE, mem);
 	/* If we are not able to push anything with `push__charbuf`
 	 * `charbuf_len` will be equal to `0` so we will send an
 	 * empty param to userspace.
 	 */
 	push__param_len(auxmap->data, &auxmap->lengths_pos, charbuf_len);
 	return charbuf_len;
 }
+
+/**
+ * @brief This helper stores the file path extracted from the `fd`.
+ *
+ * Please note: Kernel 5.10 introduced a new bpf_helper called `bpf_d_path`
+ * to extract a file path starting from a file descriptor but it can be used only
+ * with specific hooks:
+ *
+ * https://github.com/torvalds/linux/blob/e0dccc3b76fb35bb257b4118367a883073d7390e/kernel/trace/bpf_trace.c#L915-L929.
+ *
+ * So we need to do it by hand and this cause a limit in the max
+ * path component that we can retrieve (MAX_PATH_POINTERS).
+ *
+ * @param auxmap pointer to the auxmap in which we are storing the param.
+ * @param fd file descriptor from which we want to retrieve the file path.
+ */
+static __always_inline void auxmap__store_path_from_fd(struct auxiliary_map *auxmap, s32 fd)
+{
+
+	u16 total_size = 0;
+	u8 path_components = 0;
+	unsigned long path_pointers[MAX_PATH_POINTERS] = {0};
+	struct file *f = extract__file_struct_from_fd(fd);
+	struct dentry *de = BPF_CORE_READ(f, f_path.dentry);
+
+	/* We need to save the pointer to the parent because when we reach the root
+	 * the pointer will always point to itself, so we need to compare `de_p != de`.
+	 */
+	struct dentry *de_p = NULL;
+
+	for(int k = 0; k < MAX_PATH_POINTERS; ++k)
+	{
+		if(de == de_p)
+		{
+			break;
+		}
+		path_components++;
+		BPF_CORE_READ_INTO(&path_pointers[k], de, d_name.name);
+		de_p = de;
+		BPF_CORE_READ_INTO(&de, de, d_parent);
+	}
+
+	/* Reconstruct the path in reverse, using previously collected pointers.
+	 * The first component we face must be the root "/". When we read the
+	 * string in BPF with `bpf_probe_read_str()` we always add the `\0`
+	 * terminator. In this way, we will obtain something like this:
+	 *
+	 * - "/\0"
+	 * - "path_1\0"
+	 * - "path_2\0"
+	 * - "file\0"
+	 *
+	 * So putting it all together:
+	 *
+	 * 	"/\0path_1\0path_2\0file\0"
+	 *
+	 * But we want to obtain something like this:
+	 *
+	 * 	"/path_1/path_2/file\0"
+	 *
+	 * To obtain it we can replace all `\0` with `/`, but in this way we
+	 * obtain:
+	 *
+	 * 	"//path_1/path_2/file/"
+	 *
+	 * So the rationale here is if we have more than 1 `path_components`
+	 * we will erase the last path_pointer[path_components-1] = 0, and we
+	 * will directly push `/` as a single char
+	 *
+	 */
+
+	if(path_components > 1)
+	{
+		/* remove the root pointer and push only `/` without `\0`. */
+		path_pointers[(path_components - 1) & MAX_PATH_POINTERS - 1] = 0;
```

Comment:
I'd feel better about this with some more parentheses, I never remember whether & gets evaluated before - or after

Reply:
agree with you parentheses are never enough :)